pkg/postage/batchstore/reserve.go

```diff
@@ -0,0 +1,249 @@
+// - batch reserve is maximally utilised, i.e, cannot be extended and have 1-3 remain true
+package batchstore
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethersphere/bee/pkg/postage"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+// ErrBatchNotFound is returned when the postage batch is not found or expired
+var ErrBatchNotFound = errors.New("postage batch not found or expired")
+
+// DefaultDepth is the initial depth for the reserve
+const DefaultDepth = 5
+
+// Capacity = number of chunks in reserve. `2^23` was chosen to remain
+// relatively near the current 5M chunks ~25GB
+var Capacity = exp2(23)
+
+var big1 = big.NewInt(1)
+
+// reserveState records the state and is persisted in the state store
+type reserveState struct {
+	Depth    uint8    `json:"depth"`    // Radius of responsibility
+	Capacity int64    `json:"capacity"` // size of the reserve = number of chunks
+	Outer    *big.Int `json:"outer"`    // lower value limit for outer layer = the further half of chunks
+	Inner    *big.Int `json:"inner"`    // lower value limit for inner layer = the closer half of chunks
+}
+
+// unreserve is called when the batchstore decides not to reserve a batch on a PO
+// i.e.  chunk of  the batch in bins [0 upto PO] (closed  interval) are unreserved
+func (s *store) unreserve(b *postage.Batch, radius uint8) error {
+	return s.unreserveFunc(b.ID, radius)
+}
+
+// evictExpired is called when PutChainState is called (and there is 'settlement')
+func (s *store) evictExpired() error {
+	var toDelete [][]byte
+	// set until to total or inner whichever is greater
+	until := new(big.Int)
+	if s.rs.Inner.Cmp(big.NewInt(0)) > 0 && s.cs.Total.Cmp(s.rs.Inner) >= 0 {
+		until.Add(s.cs.Total, big1)
+	} else {
+		until.Set(s.rs.Inner)
+	}
+	var multiplier int64
+	err := s.store.Iterate(valueKeyPrefix, func(key, _ []byte) (bool, error) {
+		b, err := s.Get(valueKeyToID(key))
+		if err != nil {
+			return true, err
+		}
+		// terminate iteration if until is passed
+		if b.Value.Cmp(until) >= 0 {
+			return true, nil
+		}
+		//
+		if multiplier == 0 && b.Value.Cmp(s.rs.Inner) >= 0 {
+			multiplier = 1
+		}
+		//
+		if multiplier == 1 && b.Value.Cmp(s.rs.Outer) >= 0 {
+			multiplier = 2
+		}
+		// unreserve batch fully
+		err = s.unreserve(b, swarm.MaxPO)
+		if err != nil {
+			return true, err
+		}
+		s.rs.Capacity += multiplier * exp2(b.Depth-s.rs.Depth-1)
+		// if batch has no value then delete it
+		if b.Value.Cmp(s.cs.Total) <= 0 {
+			toDelete = append(toDelete, b.ID)
+		}
+		return false, nil
+	})
+	if err != nil {
+		return err
+	}
+	// set inner/outer to total if total is greater
+	s.rs.Inner.Set(until)
+	if s.rs.Outer.Cmp(until) < 0 {
+		s.rs.Outer.Set(until)
+	}
+	if err = s.store.Put(reserveStateKey, s.rs); err != nil {
+		return err
+	}
+	return s.delete(toDelete...)
+}
+
+// tier represents the sections of the reserve that can be  described as value intervals
+// 0 - out of reserve
```

Comment:
you can inline the comments within the `const` declaration below

Reply:
Ok